### PR TITLE
Add ParamConverter for Admins

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

This is something I had half way done (and it's not finished), just for comments if you think it's interesting or just to not continue.

The idea with this is that someone using `sensio/framework-extra-bundle` can create a controller like this:

```php
final class MyCustomController
{
    public function myCustomAction(MyAdmin $admin)
    {
        // ...
    }
}
```
Previously the admin has to be configured to use `MyCustomController` and add the `myCustom` route.

This could be a way to not force the user to extend `CRUDController` if we use an approach like https://github.com/sonata-project/SonataAdminBundle/pull/6796#issuecomment-764880016

Also the `AdminFetcher` (or another name) could be reuse for fetching the `Admin` in `CRUDController`.

I am targeting this branch, because these changes are BC.

## Changelog

```markdown
### Added
- Added `AdminFetcher` to fetch the `Admin` service associated to a request
- Added `AdminParamConverter` to be able to inject an Admin service in a controller
```

## To do

- [x] Add tests;
- [x] Load `param_converter.php` only when `sensio/framework-extra-bundle` is loaded;
- [x] Add docs.

### CODE REVIEW

**Comment by franmomu** [comment]
I haven't add docs yet because I'm not sure where to write them, should we change https://sonata-project.org/bundles/admin/master/doc/cookbook/recipe_custom_action.html? to remove the extending part of `CRUDController`?

With this, a user with custom routes defined like (this is not new, @wbloszyk wrote it down somewhere):

```php
protected function configureRoutes(RouteCollection $collection): void
{
    $collection->add('withAnnotation', null, [
        '_controller' => 'Sonata\AdminBundle\Tests\App\Controller\ParamConverterController::withAnnotation',
    ]);

    $collection->add('withoutAnnotation', null, [
        '_controller' => 'Sonata\AdminBundle\Tests\App\Controller\ParamConverterController::withoutAnnotation',
    ]);

    $collection->add('invokable', null, [
        '_controller' => InvokableController::class,
    ]);
}
```
And then the controller using ParamConverter:

```php
use Sonata\AdminBundle\Tests\App\Admin\TestingParamConverterAdmin;
use Symfony\Component\HttpFoundation\Response;

final class InvokableController
{
    public function __invoke(TestingParamConverterAdmin $admin): Response
    {
        return new Response();
    }
}
```
or using `AdminFetcher`:

```php
use Sonata\AdminBundle\Request\AdminFetcher;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;

final class InvokableController
{
    /**
     * @var AdminFetcher
     */
    private $adminFetcher;

    public function __construct(AdminFetcher $adminFetcher)
    {
        $this->adminFetcher = $adminFetcher;
    }

    public function __invoke(Request $request): Response
    {
        $admin = $this->adminFetcher->get($request);
        
        // ...

        return new Response();
    }
}
```

File: composer.json [hunk 1]
```diff
@@ -81,6 +82,7 @@
         "phpstan/phpstan-symfony": "^0.12.10",
         "psalm/plugin-symfony": "^2.0",
         "psr/event-dispatcher": "^1.0",
+        "sensio/framework-extra-bundle": "^6.1",
```
Comment (core23) [inline composer.json:85]: When adding a new `dev` dependency that can be used in production (no test dependency), you should add this to the conflict section.

**Review by core23 — DISMISSED** [review]

File: composer.json [hunk 2]
```diff
@@ -67,12 +67,14 @@
     },
     "conflict": {
         "doctrine/doctrine-bundle": ">=3",
+        "sensio/framework-extra-bundle": "<6.1",
```
Comment (VincentLanglet) [inline composer.json:70]: Does it really require 6.1 ? Or can it be 6.0 ? 5.x ?

**Review by VincentLanglet — COMMENTED** [review]
You removed the doc for the CRUD Controller.
But using the CRUDController does not seems to be a deprecated way to do it. Shouldn't we keep the CRUDController doc too ?

Also, the changelog need to be updated

**Comment by VincentLanglet** [comment]
> You removed the doc for the CRUD Controller.
> But using the CRUDController does not seems to be a deprecated way to do it. Shouldn't we keep the CRUDController doc too ?
> 
> Also, the changelog need to be updated

Friendly ping @franmomu

**Comment by franmomu** [comment]
> You removed the doc for the CRUD Controller.
> But using the CRUDController does not seems to be a deprecated way to do it. Shouldn't we keep the CRUDController doc too ?

I wasn't sure what to do and since the idea is to remove the constructor from `AbstractAdmin`, this could be a way to not need the last parameter (`$baseControllerName`). 

I would deprecate the last parameter in favor of this `_controller` option (in another PR), but I'm not sure if we all agree with this @sonata-project/contributors

**Comment by jordisala1991** [comment]
If we find a solution for the other Sonata Bundles, it is fine by me: https://github.com/sonata-project/SonataMediaBundle/blob/3.x/src/Controller/GalleryAdminController.php

There are of course more examples of CrudController extension across Sonata

**Comment by VincentLanglet** [comment]
CRUDController allows to create new actions, but also to override the existing ones (create/list/edit/...). Another feature I use a lot is to override the `redirectTo` method in order to redirect to another URL after an edit for example.

I'm not sure everything can be easily replaced by ParamConverter. And if it's still possible, it would require a lot of work for us in our project and for the developer in his project. So I would prefer to not start the deprecation in 3.x in order to not have a bigger 4.0.

I'm fine adding this feature in 3.x. But I would have kept both this documentation and the CRUDController documentation for now.


IMHO, after  the merge of  the  4 PR  https://github.com/sonata-project/SonataAdminBundle/pulls?q=is%3Aopen+is%3Apr+milestone%3A4.0, we could consider  an alpha-2  and  maybe the real 4.0

File: composer.json [hunk 3]
```diff
@@ -81,6 +83,7 @@
         "phpstan/phpstan-symfony": "^0.12.10",
         "psalm/plugin-symfony": "^2.0",
         "psr/event-dispatcher": "^1.0",
+        "sensio/framework-extra-bundle": "^5.6 || ^6.1",
```
Comment (jordisala1991) [inline composer.json:86]: Do we need to support 2 versions?

**Comment by franmomu** [comment]
**tl;dr**: Deprecating `$baseControllerName` doesn't mean users cannot extend `CRUDController`.

About deprecating the last parameter, it doesn't mean that users cannot extend `CRUDController`, the only change needed is [to specify the controller when adding a custom route](https://github.com/sonata-project/SonataAdminBundle/pull/6950#issuecomment-803550313), everything else would work the same way. 

For people overriding actions from `CRUDController`, I think these actions should be `final` and we should provide some extension points if needed. The natural evolution (not for `4.0` obviously) would be to extract those actions to their own file.

[The default controller](https://github.com/sonata-project/SonataAdminBundle/blob/01677b25381aa3c3d87dfaf7fe8699d9bc4744f3/src/DependencyInjection/Configuration.php#L121-L125) would be injected in https://github.com/sonata-project/SonataAdminBundle/blob/a8f16a1f4962251aa2bd816ae216d58e1b2024a0/src/Route/PathInfoBuilder.php in order to specify the controller [for those routes](https://github.com/sonata-project/SonataAdminBundle/blob/a8f16a1f4962251aa2bd816ae216d58e1b2024a0/src/Route/PathInfoBuilder.php#L37-L61).

If we don't add those extension points (for now), another solution for people overriding actions that would require less work for developers would be (haven't tried, but I guess it would work) to redefine the route, so for example for  [`GalleryAdminController`](https://github.com/sonata-project/SonataMediaBundle/blob/3.x/src/Controller/GalleryAdminController.php#L48) could be:

```php
protected function configureRoutes(RouteCollection $collection): void
{
    $collection->add('list', null, [
        '_controller' => 'Sonata\MediaBundle\Controller\GalleryAdminController::listAction',
    ]);
}
```

PS: The ParamConverter is only new feature for the ones who don't want to extend `CRUDController`.

File: composer.json (hunk 3 above)
Comment (franmomu) [reply composer.json:86]: I used first only `6.1`, but @VincentLanglet suggested to support more https://github.com/sonata-project/SonataAdminBundle/pull/6950#pullrequestreview-617961191

**Comment by VincentLanglet** [comment]
> For people overriding actions from `CRUDController`, I think these actions should be `final` and we should provide some extension points if needed. The natural evolution (not for `4.0` obviously) would be to extract those actions to their own file.

I took a look about how we override some route in our project.
- `createAction` is sometimes overriden to redirect to another route before the create in some special cases. I guess it can be done with the `preCreate`.
- we override a lot the `redirectTo`, it will still be available.
- we add some batchActions

So I'm fine making them final indeed.
And what would be the strategy for the batchActions ? https://symfony.com/doc/current/bundles/SonataAdminBundle/reference/batch_actions.html

> About deprecating the last parameter, it doesn't mean that users cannot extend `CRUDController`, the only change needed is [to specify the controller when adding a custom route](https://github.com/sonata-project/SonataAdminBundle/pull/6950#issuecomment-803550313), everything else would work the same way.
> [The default controller](https://github.com/sonata-project/SonataAdminBundle/blob/01677b25381aa3c3d87dfaf7fe8699d9bc4744f3/src/DependencyInjection/Configuration.php#L121-L125) would be injected in https://github.com/sonata-project/SonataAdminBundle/blob/a8f16a1f4962251aa2bd816ae216d58e1b2024a0/src/Route/PathInfoBuilder.php in order to specify the controller [for those routes](https://github.com/sonata-project/SonataAdminBundle/blob/a8f16a1f4962251aa2bd816ae216d58e1b2024a0/src/Route/PathInfoBuilder.php#L37-L61).

If you don't provide the CRUDController as third parameter of the admin, how your admin is link to a custom CRUDController when you're overriding some extension point like `redirectTo` ?

> PS: The ParamConverter is only new feature for the ones who don't want to extend `CRUDController`.

Sure, that's why I have no blocker for merging the new feature in 3.x.

File: composer.json (hunk 3 above)
Comment (jordisala1991) [reply composer.json:86]: At least for master we should remove ^5 support

**Comment by franmomu** [comment]
Hmm maybe as a first step it's easier to just add a `setBaseControllerName` and deprecate `$baseControllerName`, but that's another PR.

File: composer.json (hunk 3 above)
Comment (VincentLanglet) [reply composer.json:86]: I'm not against the removal of the `^5` support. 
My review was more about the conflict https://github.com/sonata-project/SonataAdminBundle/pull/6950#pullrequestreview-617961191

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
> Hmm maybe as a first step it's easier to just add a `setBaseControllerName` and deprecate `$baseControllerName`, but that's another PR.

Can you create an issue about the whole `$baseControllerName` deprecation strategy ? :)

File: composer.json (hunk 3 above)
Comment (jordisala1991) [reply composer.json:86]: Oh I see, but it is nice conflict matches require dev, in order to make sure our code works with the installed version of the user project

Comment (jordisala1991) [reply composer.json:86]: Can we use only 6.1 then? @VincentLanglet and @franmomu ?

Comment (VincentLanglet) [reply composer.json:86]: Sure we can remove the test of 5.6 if you want, but the conflict should stay on 5.6 since it works with upper version

Comment (jordisala1991) [reply composer.json:86]: Then dont change anything it does not make sense to conflict with different thing that you are installing :/

Comment (VincentLanglet) [reply composer.json:86]: > it is nice conflict matches require dev,

Conflict are here to say `This cannot works for sure with these versions`.
The feature required for this PR was introduced in the version 5.6.

Having test on all the possible version is another thing.

Comment (franmomu) [reply composer.json:86]: If we keep `5.6` in the `conflict` section, to me it doesn't make sense to not test it as well, let's keep it this way and remove it after merging?

**Review by jordisala1991 — APPROVED** [review]
